packages/core/models/resources/lib/defaultValue.ts

```diff
@@ -1,27 +1,61 @@
-import type { PropertyShape } from '@rdfine/shacl'
+import type { NodeKind, PropertyShape } from '@rdfine/shacl'
 import type { GraphPointer, MultiPointer } from 'clownface'
 import { dash, rdf, sh } from '@tpluscode/rdf-ns-builders'
 import type { ResourceIdentifier } from '@tpluscode/rdfine'
+import type { NamedNode } from 'rdf-js'
+import { nanoid } from 'nanoid'
 import type { FocusNode } from '../../../index'
 
 export function defaultValue(property: PropertyShape, focusNode: FocusNode): MultiPointer | null {
-  const { nodeKind } = property
-
   if (property.defaultValue) {
     return focusNode.node(property.defaultValue)
   }
 
+  const { nodeKind } = property
+  switch (nodeKind?.value) {
+    case 'http://www.w3.org/ns/shacl#IRI':
+    case 'http://www.w3.org/ns/shacl#IRIOrLiteral':
+    case 'http://www.w3.org/ns/shacl#BlankNode':
+    case 'http://www.w3.org/ns/shacl#BlankNodeOrIRI':
+    case 'http://www.w3.org/ns/shacl#BlankNodeOrLiteral':
+      return createResourceNode(property, nodeKind, focusNode)
+    default:
+      return null
+  }
+}
+
+function createResourceNode(property: PropertyShape, nodeKind: NodeKind, focusNode: FocusNode) {
+  const uriStart = property.pointer.out(dash.uriStart).value
+  let resourceNode: GraphPointer<ResourceIdentifier> = focusNode.blankNode()
+
+  if (nodeKind.equals(sh.IRI)) {
+    resourceNode = createNewUri(property, focusNode, uriStart)
+  } else if (uriStart && nodeKind.equals(sh.BlankNodeOrIRI)) {
+    resourceNode = createNewUri(property, focusNode, uriStart)
+  } else if (nodeKind.equals(sh.IRIOrLiteral)) {
+    if (!uriStart) {
+      return null
+    }
+
+    resourceNode = createNewUri(property, focusNode, uriStart)
+  }
+
   const propertyClass = property.class
-  if (propertyClass || nodeKind?.equals(sh.IRI) || nodeKind?.equals(sh.BlankNode) || nodeKind?.equals(sh.BlankNodeOrIRI)) {
-    const resourceNode: GraphPointer<ResourceIdentifier> = nodeKind?.equals(sh.IRI) ? focusNode.namedNode('') : focusNode.blankNode()
+  if (propertyClass || nodeKind.equals(sh.IRI) || nodeKind.equals(sh.BlankNode) || nodeKind.equals(sh.BlankNodeOrIRI)) {
     if (propertyClass && !property.get(dash.editor)?.equals(dash.InstancesSelectEditor)) {
       return resourceNode.addOut(rdf.type, propertyClass.id)
     }
   }
 
-  if (nodeKind?.equals(sh.BlankNode) || nodeKind?.equals(sh.BlankNodeOrIRI)) {
-    return focusNode.blankNode()
+  return resourceNode
+}
+
+function createNewUri(property: PropertyShape, focusNode: FocusNode, uriStart: string | undefined): GraphPointer<NamedNode> {
```

Comment:
I would add an options{} object to the `default` function and put this function as an optional parameter. 

Just in case one doesn't want to use nanoid() to mint URIs (for use, for example, use incremental integers instead).

The default would be nanoid()

Reply:
That is a very good idea although I'm uncertain about the details until we come across a need for that.

**level of granularity** 

Should the extension point be on the function which creates the URI component (what nano does here), or maybe the injected function would return the while URI reference (possible ignoring `sh1:iriPrefix` completely

**Dependency Injection**

How would the function be injected?
Using triples, by referencing code in the Shapes Graph? 
Or purely in code, through the Shaperone package somewhere?

Reply:
I was thinking about a function passed in the options, nothing fancy. 
But you are right, we haven't come across that